core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -847,6 +847,301 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::TeamThreadRangeBoundariesStruct<
+
+enum class MDTeamRangeEndRank : bool { NotEndRank, EndRank };
+enum class MDTeamRangeParThread : bool { NotParThread, ParThread };
+enum class MDTeamRangeParVector : bool { NotParVector, ParVector };
+enum class MDTeamRangeThreadAndVector : bool { NotBoth, Both };
+
+struct NoReductionTag {
+  explicit NoReductionTag() = default;
+};
+
+inline constexpr NoReductionTag no_reduction_tag{};
+
+// Tag class to choose the nested loop specialization
+//   - EndRank means call the actual closure
+//   - ParThread means use TeamThreadRange
+//   - ParVector means use ThreadVectorRange
+template <MDTeamRangeEndRank EndRank, MDTeamRangeParThread ParThread,
+          MDTeamRangeParVector ParVector>
+struct MDTeamRangeMode {
+  static constexpr MDTeamRangeEndRank end_rank     = EndRank;
+  static constexpr MDTeamRangeParThread par_thread = ParThread;
+  static constexpr MDTeamRangeParVector par_vector = ParVector;
+};
+
+// Tag class to keep track of the loop nest level and where to deploy thread and
+// vector parallelism
+//   - Rank is Kokkos::Rank<TotalRank, Iter>
+//     - TotalRank is the total number of loop nests
+//     - Iter is whether to go forward or backward through ranks (i.e. the
+//       iteration order for MDRangePolicy)
+//   - ParThreadRank is the nesting level on which to deploy thread parallelism
+//   - ParVectorRank is the nesting level on which to deploy vector parallelism
+//   - CurrentRank is the current nest level
+template <typename Rank, int ParThreadRank, int ParVectorRank, int CurrentRank>
+struct MDTeamRangeRanks {
+  using rank_type                            = int;
+  static constexpr Iterate iter              = Rank::outer_direction;
+  static constexpr rank_type total_rank      = Rank::rank;
+  static constexpr rank_type par_thread_rank = ParThreadRank;
+  static constexpr rank_type par_vector_rank = ParVectorRank;
+  static constexpr rank_type current_rank    = CurrentRank;
+
+  // We have to recursively process ranks [0..total_rank-1]
+  using RangeMode = MDTeamRangeMode<
+      (iter == Iterate::Right)
+          ? static_cast<MDTeamRangeEndRank>(current_rank == total_rank)
+          : static_cast<MDTeamRangeEndRank>(current_rank == -1),
+      static_cast<MDTeamRangeParThread>(current_rank == par_thread_rank),
+      static_cast<MDTeamRangeParVector>(current_rank == par_vector_rank)>;
+};
+
+// ParallelRank determines on which rank parallelization happens.
+//   - Rank is Kokkos::Rank<TotalRank, Iter>
+//     - TotalRank is the total number of loop nests
+//     - Iter is whether to go forward or backward through ranks (i.e. the
+//       iteration order for MDRangePolicy)
+//   - t_and_v determines whether both vector and thread parallelism is in use
+template <typename Rank, typename ExecSpace, MDTeamRangeThreadAndVector t_and_v>
+struct ParallelRank;
+
+#ifdef KOKKOS_ENABLE_SERIAL
+template <typename Rank, MDTeamRangeThreadAndVector t_and_v>
+struct ParallelRank<Rank, DefaultHostExecutionSpace, t_and_v> {
```

Comment:
```suggestion
struct ParallelRank<Rank, Serial, t_and_v> {
```

Reply:
Changed in [7361243](https://github.com/kokkos/kokkos/pull/5238/commits/7361243c9f3e76c1328b1c1d27df3f25f0ec410a).